src/main/java/org/nebula/contrib/ngbatis/proxy/MapperProxy.java

```diff
@@ -236,23 +241,31 @@ public static ResultSet executeWithParameter(ClassModel cm, MethodModel mm, Stri
       if (log.isDebugEnabled()) {
         log.debug("\n\t- proxyMethod: {}#{}"
                 + "\n\t- session space: {}"
+                + (isEmpty(autoSwitch) ? "{}" : "\n\t- auto switch to: {}")
                 + "\n\t- nGql：{}"
                 + "\n\t- params: {}"
                 + "\n\t- result：{}",
-            proxyClass, proxyMethod, localSessionSpace, gql, params, result);
+            proxyClass, proxyMethod, localSessionSpace, autoSwitch, gql, params, result);
       }
       if (localSession != null) {
         ENV.getDispatcher().offer(localSession);
       }
     }
   }
 
-  private static String qlWithSpace(LocalSession localSession, String gql, String currentSpace) {
+  private static String[] qlWithSpace(LocalSession localSession, String gql, String currentSpace)
+      throws IOErrorException {
+    String[] qlAndSpace = new String[2];
     gql = gql.trim();
     String sessionSpace = localSession.getCurrentSpace();
-    return Objects.equals(sessionSpace, currentSpace)
-        ? String.format("\n\t\t%s", gql)
-        : String.format("USE %s;\n\t\t%s", currentSpace, gql);
+    boolean sameSpace = Objects.equals(sessionSpace, currentSpace);
+    if (!sameSpace) {
+      qlAndSpace[0] = currentSpace;
+      Session session = localSession.getSession();
+      session.execute(String.format("USE %s", currentSpace));

```

Comment:
I prefer to execute `USE space` after each request.  In resultSet of execution's response, there's space name for current session, and we can decide whether need to change the space back.

If execute `USE space ` before request, there will be some useless operations, ge:
Local session's current space is test, and user's request is "use test1; show hosts; use test" , and  there is no need to pre-execute a `use test1` operation.

Reply:
In this case, you are right.

And we should consider another situation, for example there are two DAO: 
```xml
<mapper namespace="Test1Dao" space="test1">

    <select id="useCurrentSpace" >
         RETURN 'useCurrentSpace';
    </select>

    <select id="useTestSpace" space="test">
         RETURN 'useTestSpace';
    </select>

    <select id="useSpaceInner">
         USE test;
         RETURN 'useSpaceInner';
    </select>
    
</mapper>


<mapper namespace="DefaultSpaceInYmlDao"> <!-- default space is `test`-->
    
    <select id="normalInvoke">
         RETURN 'normalInvoke';
    </select>

</mapper>
```

```java
    @Autowired private Test1Dao test1Dao;
    @Autowired private DefaultSpaceInYmlDao defaultSpaceDao;
    
    public void biz() {
        test1Dao.useCurrentSpace(); 
        // USE test1;
        // RETURN 'useCurrentSpace';
        test1Dao.useCurrentSpace(); // local session keep the last space, so this time is unnecessary to switch
        // RETURN 'useCurrentSpace';
        test1Dao.useTestSpace(); // catching the space is difference from local session.
        // USE test;
        // RETURN 'useTestSpace';
        defaultSpaceDao.normalInvoke(); 
        // RETURN 'normalInvoke';
        test1Dao.useTestSpace();
        // RETURN 'useTestSpace';
        test1Dao.useSpaceInner(); // This usage #78 may happend again, because we can not control how ngql will be written
        // USE test; RETURN 'useSpaceInner'; 
    }
```

Let's assume that only one session is in use.
In this case 6 times invoked, we costs 8 times request. not 12 times.

Ngbatis recommends declaring spaces in xml tags.

And you reminded me that we can get current session space from ResultSet insteads of RegExp.

Reply:
In this case, you are right.

And we should consider another situation, for example there are two DAO: 
```xml
<mapper namespace="Test1Dao" space="test1">

    <select id="useCurrentSpace" >
         RETURN 'useCurrentSpace';
    </select>

    <select id="useTestSpace" space="test">
         RETURN 'useTestSpace';
    </select>

    <select id="useSpaceInner">
         USE test;
         RETURN 'useSpaceInner';
    </select>
    
</mapper>


<mapper namespace="DefaultSpaceInYmlDao"> <!-- default space is `test`-->
    
    <select id="normalInvoke">
         RETURN 'normalInvoke';
    </select>

</mapper>
```

```java
    @Autowired private Test1Dao test1Dao;
    @Autowired private DefaultSpaceInYmlDao defaultSpaceDao;
    
    public void biz() {
        test1Dao.useCurrentSpace(); 
        // USE test1;
        // RETURN 'useCurrentSpace';
        test1Dao.useCurrentSpace(); // local session keep the last space, so this time is unnecessary to switch
        // RETURN 'useCurrentSpace';
        test1Dao.useTestSpace(); // catching the space is difference from local session.
        // USE test;
        // RETURN 'useTestSpace';
        defaultSpaceDao.normalInvoke(); 
        // RETURN 'normalInvoke';
        test1Dao.useTestSpace();
        // RETURN 'useTestSpace';
        test1Dao.useSpaceInner(); // This usage #78 may happend again, because we can not control how ngql will be written
        // USE test; RETURN 'useSpaceInner'; 
    }
```

Let's assume that only one session is in use.
In this case 6 times invoked, we costs 8 times request. not 12 times.

Ngbatis recommends declaring spaces in xml tags.

And you reminded me that we can get current session space from ResultSet insteads of RegExp.